src/Effect/Aff.purs

```diff
@@ -101,185 +103,195 @@ instance monadRecAff ∷ MonadRec Aff where
 
 -- | Forks an `Aff` from an `Effect` context, returning the `Fiber`.
-launchAff ∷ ∀ a. Aff a → Effect (Fiber a)
+launchAff ∷ ∀ e a. Aff e a → Effect (Fiber e a)
 launchAff aff = do
   fiber ← makeFiber aff
   case fiber of Fiber f → f.run
   pure fiber
 
 -- | Forks an `Aff` from an `Effect` context, discarding the `Fiber`.
-launchAff_ ∷ ∀ a. Aff a → Effect Unit
+launchAff_ ∷ ∀ e a. Aff e a → Effect Unit
 launchAff_ = void <<< launchAff
 
 -- | Suspends an `Aff` from an `Effect` context, returning the `Fiber`.
-launchSuspendedAff ∷ ∀ a. Aff a → Effect (Fiber a)
+launchSuspendedAff ∷ ∀ e a. Aff e a → Effect (Fiber e a)
 launchSuspendedAff = makeFiber
 
 -- | Forks an `Aff` from an `Effect` context and also takes a callback to run when
 -- | it completes. Returns the pending `Fiber`.
-runAff ∷ ∀ a. (Either Error a → Effect Unit) → Aff a → Effect (Fiber Unit)
+runAff ∷ ∀ e a. (Either e a → Effect Unit) → Aff e a → Effect (Fiber e Unit)
 runAff k aff = launchAff $ liftEffect <<< k =<< try aff
 
 -- | Forks an `Aff` from an `Effect` context and also takes a callback to run when
 -- | it completes, discarding the `Fiber`.
-runAff_ ∷ ∀ a. (Either Error a → Effect Unit) → Aff a → Effect Unit
+runAff_ ∷ ∀ e a. (Either e a → Effect Unit) → Aff e a → Effect Unit
 runAff_ k aff = void $ runAff k aff
 
 -- | Suspends an `Aff` from an `Effect` context and also takes a callback to run
 -- | when it completes. Returns the suspended `Fiber`.
-runSuspendedAff ∷ ∀ a. (Either Error a → Effect Unit) → Aff a → Effect (Fiber Unit)
+runSuspendedAff ∷ ∀ a. (Either Unit a → Effect Unit) → Aff Unit a → Effect (Fiber Unit Unit)
 runSuspendedAff k aff = launchSuspendedAff $ liftEffect <<< k =<< try aff
 
 -- | Forks am `Aff` from within a parent `Aff` context, returning the `Fiber`.
-forkAff ∷ ∀ a. Aff a → Aff (Fiber a)
+forkAff ∷ ∀ e1 e2 a. Aff e1 a → Aff e2 (Fiber e1 a)
 forkAff = _fork true
 
 -- | Suspends an `Aff` from within a parent `Aff` context, returning the `Fiber`.
 -- | A suspended `Aff` is not executed until a consumer observes the result
 -- | with `joinFiber`.
-suspendAff ∷ ∀ a. Aff a → Aff (Fiber a)
+suspendAff ∷ ∀ e1 e2 a. Aff e1 a → Aff e2 (Fiber e1 a)
 suspendAff = _fork false
 
 -- | Pauses the running fiber.
-delay ∷ Milliseconds → Aff Unit
+delay ∷ ∀ e. Milliseconds → Aff e Unit
 delay (Milliseconds n) = Fn.runFn2 _delay Right n
 
 -- | An async computation which does not resolve.
-never ∷ ∀ a. Aff a
+never ∷ ∀ e a. Aff e a
 never = makeAff \_ → pure mempty
 
--- | A monomorphic version of `try`. Catches thrown errors and lifts them
--- | into an `Either`.
-attempt ∷ ∀ a. Aff a → Aff (Either Error a)
-attempt = try
+-- | A version of `catchError` that can map the error type.
+catch ∷ ∀ e1 e2 a. Aff e1 a → (e1 → Aff e2 a) → Aff e2 a
+catch = _catchError
+
+-- | A monomorphic version of `try` that can map the error type. Catches thrown
+-- | errors and lifts them into an `Either`.
+attempt ∷ ∀ e1 e2 a. Aff e1 a → Aff e2 (Either e1 a)
```

Comment:
There should be `Void` instead of `e2` i suppose

Reply:
`Aff e a` does not mean an error of type `e` is _definitely_ thrown; it means an error of type `e` _may_ be thrown.

Reply:
Yes but if you do `attempt` on some Aff `computation` it must eliminate all possible errors, which means resulting Aff must throw nothing. If you stick `Void` there it would mean `Void may be throw` but as there is no value of type Void it means nothing can be thrown.

Reply:
That would prevent you from `attempt`ing inside an exiting `Aff e1`, unless you `lmap absurd (attempt ...)` though, which would be a bit annoying.

Reply:
good point